### DISCUSSION TITLE
Inverter power limit for low solar input

### DISCUSSION THREAD

**Discussion by ahiggs1** [hoylabs/OpenDTU-OnBattery#1459] [General]
It would be good if you could set the inverter power limit to equal the MPPT total input power when the battery SoC is at/near the _discharge start threshold_. 

The _battery current limit_ below X% SoC setting almost fills this function. However this limit is applied to the DC input current of the inverter and not the actual battery discharge rate. Maybe this is a bug?

For example, consider the following scenario:
400W MPPT Total power
24% battery SoC, 50V battery voltage, 20% _discharge start threshold_, 15% _discharge stop threshold_
5A _battery current limit_ set for SoC <25%
600W grid consumption

The optimal control would be to set the inverter power limit to 400W and keep the battery SoC constant around the _discharge stop threshold_. However what actually happens with these settings is the inverter power is set to ~250W and the battery would charge by ~150W

**Comment by schlimmchen** [comment]
> It would be good if you could set the inverter power limit to equal the MPPT total input power when the battery SoC is at/near the discharge start threshold.

You are asking for **full** solar-passthrough? If you want the DPL to set the battery-powered inverter(s) to invert all solar charge controller input power to AC, you can set the **full** solar-passthrough setting to the same value or slightly above the start threshold. Not sure why you would want that? Maybe you want to use the power somewhere else as soon as the battery is "mildly" charged?

> The battery current limit below X% SoC setting almost fills this function.

Does it? Not sure that I am understanding correctly...

> However this limit is applied to the DC input current of the inverter and not the actual battery discharge rate. Maybe this is a bug?

The battery discharge limit is calculated based on your settings, but possibly also taking the discharge limit into account that your BMS is reporting. That discharge limit is interpreted as an allowance in the DPL. The allowance is increased by the solar input from the charge controller, if solar-passthrough is enabled. The inverter's DC input can be higher than the battery discharge allowance if there is power from the solar charger.

> However what actually happens with these settings is the inverter power is set to ~250W and the battery would charge by ~150W

I am confused...

Are you on the latest release and or on a build from the development branch?
What is your **full** solar-passthrough start threshold?
You are using a VE.Direct solar charger which is connected to OpenDTU-OnBattery?

It would be helpful if you could provide a verbose log from when the issue is present.

**Reply by ahiggs1** [reply]
For days when it is cloudy and the solar input is below the current grid load, I want the inverter power limit to equal the solar input power when the battery is empty. 

> The inverter's DC input can be higher than the battery discharge allowance if there is power from the solar charger.

This is not what I experienced when I tested this setting. I will check again this weekend when I am home in the morning and can reproduce the scenario. 

> Are you on the latest release and or on a build from the development branch?
> What is your full solar-passthrough start threshold?
> You are using a VE.Direct solar charger which is connected to OpenDTU-OnBattery?

Latest release, full solar-passthrough disabled. Both Smart solar MPPTs are connected with VE.Direct and the Pytes V5a with CAN, never had any communication issues

**Reply by ahiggs1** [reply]
[BatteryDischargeConfig.json](https://github.com/user-attachments/files/18218755/BatteryDischargeConfig.json)
[BatteryDischargeLimitLog.txt](https://github.com/user-attachments/files/18218756/BatteryDischargeLimitLog.txt)
![Screenshot 2024-12-21 084639](https://github.com/user-attachments/assets/fc1fd775-f40e-46e1-a3cd-f757ec322fd6)


Ok, so I managed to take the log today in the morning. As shown in the screenshot, MPPT power is 420W, grid power is 263W, but the battery is charging with 312W and inverter power is 86W. The battery start discharge threshold has been passed (set to 16%), but the current limiter is set to 2A below 25%. Is there something incorrect in the settings?

**Reply by AndreasBoehm** [reply]
In the config file I can see that you disabled "solar passthrough" this means that solar power will be ignored and we only rely on power available from the battery. This explains why a lot of power goes into the battery but only fraction goes to the inverter.

The remaining question is why you can only draw 85 watts from the battery.

**Reply by schlimmchen** [reply]
> [DPL::loop] StartTH reached: yes, StopTH reached: no, **SolarPT disabled**, use at night: yes

If you want to make use of the power the solar charger contribute to the "power bus", you'll need to enable solar-passthrough. If you did, the total power available from the power bus is 420 + 95 W. Withouth solar-passthrough, it is only 95W as you limit battery discharging to 2A.

I can see how that is weird, but the way the DPL is implemented right now assumes that the power drawn from the "power bus" is drawn from the battery if solar-passthrough is off. The DPL pretends to not know the solar charger's output. Hm... not sure if it makes sense to have it this way. Even if the solar charge controller's output is actually unknown, but the battery discharge or charge rate is known (from the BMS), we could still draw power from the power bus until the battery is actually drained at the requested 2A. Only if neither is known, we must conservatively limit the inverter(s) to the battery discharge limit as we must assume the worst case which is that all power is coming from the battery...

Well, in your case: Can you enable solar-passthrough or do you want it off for some reason?

**Reply by schlimmchen** [reply]
> The remaining question is why you can only draw 85 watts from the battery.

Nah, that's easy :wink: The discharge limit is 2A, and the battery voltage is 51.94V. That's 104W. With cable losses and inverter efficiency we are at 95W, which is part of the log as well. As the inverter already produces 85W but the hysteresis is 20W, no action is taken.

**Reply by ahiggs1** [reply]
Ok, thanks for the clarification. I didn't realize that there was a difference between solar pass-through and full solar pass-through so I never enabled it as I don't really want to feed the grid. I'll give this a shot and see if it's better tomorrow (assuming it's sunny...)

**Reply by SW-Niko** [reply]
> The optimal control would be to set the inverter power limit to 400W and keep the battery SoC constant around the discharge stop threshold. However what actually happens with these settings is the inverter power is set to ~250W and the battery would charge by ~150W

Hint: One side effect of #1111 is the behavior described above. I really have to finish my work ....
